File: hooks/stop-nixos-rebuild-on-config-change.py
#!/usr/bin/env python3
"""Stop hook: rebuild NixOS when nix-config HEAD has changed since last rebuild."""

import json
import subprocess
import sys
import tempfile
from pathlib import Path

HASH_FILE = Path(tempfile.gettempdir()) / ".nix-config-last-rebuild-hash"
NIX_CONFIG = Path.home() / "nix-config"


def run(*args: str, cwd: Path = NIX_CONFIG) -> subprocess.CompletedProcess[str]:
    return subprocess.run(args, capture_output=True, text=True, cwd=cwd)


def main() -> None:
    data = json.load(sys.stdin)

    if data.get("permission_mode") == "plan":
        return

    # tree が dirty なら rebuild しない（stop-git-check.py に任せる）
    result = run("git", "status", "--porcelain")
    if result.stdout.strip():
        return

    # 現在の HEAD ハッシュを取得
    result = run("git", "rev-parse", "HEAD")
    if result.returncode != 0:
        return
    current_hash = result.stdout.strip()

    # 前回リビルド時のハッシュと比較
    last_hash = HASH_FILE.read_text().strip() if HASH_FILE.exists() else ""
    if current_hash == last_hash:
        return

    result = run("sudo", "nixos-rebuild", "switch", "--flake", f"{NIX_CONFIG}#nixos")
    if result.returncode != 0:
        json.dump(
            {"decision": "block", "reason": f"nixos-rebuild failed:\n{result.stderr}"},
            sys.stdout,
        )
    else:
        HASH_FILE.write_text(current_hash)


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/stop_nixos_rebuild_on_config_change.rs
use crate::git::run_git;
use crate::input::HookInput;
use crate::output::block;

fn should_rebuild(status: &str, head_hash: &str, last_hash: &str) -> bool {
    status.is_empty() && !head_hash.is_empty() && head_hash != last_hash
}

pub fn run(input: &HookInput) {
    if input.permission_mode == "plan" {
        return;
    }

    let nix_config = home::home_dir().unwrap_or_default().join("nix-config");
    let status = run_git(&nix_config, &["status", "--porcelain"]);
    if !status.is_empty() {
        return;
    }

    let head_hash = run_git(&nix_config, &["rev-parse", "HEAD"]);
    if head_hash.is_empty() {
        return;
    }

    let hash_file = std::env::temp_dir().join(".nix-config-last-rebuild-hash");
    let last_hash = std::fs::read_to_string(&hash_file)
        .unwrap_or_default()
        .trim()
        .to_string();
    if !should_rebuild(&status, &head_hash, &last_hash) {
        return;
    }

    let result = std::process::Command::new("sudo")
        .args(["nixos-rebuild", "switch", "--flake"])
        .arg(format!("{}#nixos", nix_config.display()))
        .output();

    match result {
        Ok(o) if o.status.success() => {
            let _ = std::fs::write(&hash_file, &head_hash);
        }
        Ok(o) => {
            let stderr = String::from_utf8_lossy(&o.stderr).to_string();
            block(&format!("nixos-rebuild failed:\n{stderr}"));
        }
        Err(_) => {}
    }
}

#[cfg(test)]
mod tests {
    use super::should_rebuild;

    #[test]
    fn dirty_tree_skips_rebuild() {
        assert!(!should_rebuild(" M file.nix", "new_hash", "old_hash"));
    }

    #[test]
    fn missing_head_hash_skips_rebuild() {
        assert!(!should_rebuild("", "", "old_hash"));
    }

    #[test]
    fn unchanged_hash_skips_rebuild() {
        assert!(!should_rebuild("", "same_hash", "same_hash"));
    }

    #[test]
    fn clean_changed_head_triggers_rebuild() {
        assert!(should_rebuild("", "new_hash", "old_hash"));
    }
}
